tests/tests_pytorch/trainer/test_dataloaders.py

```diff
@@ -996,17 +996,18 @@ def val_dataloader(self):
         assert train_reload_epochs == [0, 1, 2, 3, 4]
         assert val_reload_epochs == [0, 1, 2, 3, 4]
     elif n == 2:
-        assert train_reload_epochs == [0, 2, 4]
-        assert val_reload_epochs == [0, 2, 4]
+        assert train_reload_epochs == [0, 1, 3]
+        assert val_reload_epochs == [0, 1, 3]
```

Comment:
I'm a bit confused here. Shouldn't the sanity check also be seen in this list here, as an additional epoch 0? Perhaps we don't reload them when sanity check ran, and we just reuse them for the first real epoch. But then I don't understand why  [0, 2, 4] wasn't correct. My reasoning: Reloading every 2 epochs means skip every other. Since loading on epoch idx 0 is mandatory, we need to skip epoch idx 1 and then in epoch idx 2 we load again, and so on.

Reply:
> Perhaps we don't reload them when sanity check ran, and we just reuse them for the first real epoch.

yes.


I had a quick discussion with @carmocca offline regarding this.

Also: https://github.com/Lightning-AI/lightning/issues/12145#issuecomment-1196912756

but here it depends upon `check_val_every_n_epoch`. By default, it's `1`. But if you set `check_val_every_n_epoch=reload_dataloaders_every_n_epochs`, it will run at the 2nd, 4th epoch.. etc. The first `0` only runs the sanity check if `check_val_every_n_epoch=reload_dataloaders_every_n_epochs`. And that's how we have integrated other flags as well. For eg. `log_every_n_steps=2`, we log at 2nd, 4th step... etc.

Let me simplify the test and fix one case.

Reply:
I have made some updates and improved the tests to check the edge cases. The new logic aligns with your suggestion and also looks most optimal to me.

Reply:
but now it does not align with the rule other arguments like `log_every_n_steps` or `val_check_interval`.